### PR TITLE
fix Issue 22842 - importC: cannot declare function with a typedef

### PR DESCRIPTION
**Pull request by WalterBright**
took me a while to figure out how to make this work

### CODE REVIEW

File: src/dmd/cparse.d [hunk 1]
```diff
@@ -4557,6 +4571,35 @@ final class CParser(AST) : Parser!AST
...
 
+    /********************************
+     * Determines if type t is a function type.
+     * Make this work without needing semantic analysis.
+     * Params:
+     *  t = type to test
+     * Returns:
+     *  true if it represents a function
+     */
+    bool isfunction(AST.Type t)
```
Comment (ibuclaw) [inline src/dmd/cparse.d:4582]: Would prefer camel-casing the function name, but not a big deal.  Also, `private`?

File: src/dmd/cparse.d [hunk 2]
```diff
@@ -4557,6 +4571,35 @@ final class CParser(AST) : Parser!AST
...
+    {
+        //printf("isfunction() %s\n", t.toChars());
+        if (t.isTypeFunction())
+            return true;
+        if (auto tid = t.isTypeIdentifier())
+        {
+            /* Scan array of typedef identifiers that are an alias for
+             * a function type
+             */
+            foreach (ftid; funcTypeIds[])
```
Comment (ibuclaw) [inline src/dmd/cparse.d:4592]: Not great that this is linear, so we're hoping that no one is going to have thousands of these typedefs in a library.

Comment (WalterBright) [reply src/dmd/cparse.d:4592]: They'd have to be thousands of _function_ typedefs, not just typedefs! Anyhow, I wanted this PR to be as simple as I could. I wanted to prove my idea for fixing this problem was workable, as it took me a couple days to come up with it. If we need a more sophisticated symbol table, that should be separate.

File: src/dmd/cparse.d (hunk 1 above)
Comment (WalterBright) [reply src/dmd/cparse.d]: done

File: src/dmd/cparse.d (hunk 2 above)
Comment (ibuclaw) [reply src/dmd/cparse.d:4592]: Actually, keeping all typedefs in scope in a symbol table might help us with some of the ugly ambiguous cases that we try to patch up in the semantic.  But this is fine as-is for now.

**Review by ibuclaw — APPROVED** [review]

**Pull request merged**